### PR TITLE
tc009xlvc.cpp : Updates

### PR DESCRIPTION
**Pull request by cam900**
Internalize on-chip functions (ROM / RAM bankswitching, GFX, IRQ vector and enables, Palette, Vregs), Convert arrays into required_shared_ptrs, Use address_map_bank_device for ram banks, Simplify handlers, Reduce unnecessary values, Lines, Pointers, Duplicates, Fix spacings, Typo, Use shorter / correct type values, Add notes
dfruit.cpp, lastbank.cpp : Reduce unnecessary address map
lastbank.cpp : Correct audiocpu ROM region size

### CODE REVIEW

File: src/devices/machine/tc009xlvc.cpp [hunk 1]
```diff
@@ -14,186 +14,133 @@
...
+	return m_rom[offset & m_rom.mask()];
 }
 
-WRITE8_MEMBER(tc0091lvc_device::tc0091lvc_paletteram_w)
+void tc0091lvc_device::vram_w(offs_t offset, u8 data)
 {
-	m_palette_ram[offset & 0x1ff] = data;
+	// note, the way tiles are addressed suggests that 0x0000-0x3fff of this might be usable,
+	//       but we don't map it anywhere, so the first tiles are always blank at the moment.
+	if (offset < 0x4000) // <0x4000 area is not mapped?
```
Comment (angelosa) [inline src/devices/machine/tc009xlvc.cpp]: Accessing like this means that the VRAM internal space needs to be routed in a different address_space, namely for a pratical reason, every time you're accessing like this you're making a new check, giving a performance penalty.

File: src/mame/drivers/dfruit.cpp [hunk 2]
```diff
@@ -103,95 +74,11 @@ WRITE_LINE_MEMBER(dfruit_state::screen_vblank)
...
-	map(0xc000, 0xcfff).rw(FUNC(dfruit_state::dfruit_ram_0_r), FUNC(dfruit_state::dfruit_ram_0_w));
-	map(0xd000, 0xdfff).rw(FUNC(dfruit_state::dfruit_ram_1_r), FUNC(dfruit_state::dfruit_ram_1_w));
-	map(0xe000, 0xefff).rw(FUNC(dfruit_state::dfruit_ram_2_r), FUNC(dfruit_state::dfruit_ram_2_w));
-	map(0xf000, 0xfdff).rw(FUNC(dfruit_state::dfruit_ram_3_r), FUNC(dfruit_state::dfruit_ram_3_w));
-
-	map(0xfe00, 0xfeff).rw(m_vdp, FUNC(tc0091lvc_device::vregs_r), FUNC(tc0091lvc_device::vregs_w));
-	map(0xff00, 0xff02).rw(FUNC(dfruit_state::dfruit_irq_vector_r), FUNC(dfruit_state::dfruit_irq_vector_w));
-	map(0xff03, 0xff03).rw(FUNC(dfruit_state::dfruit_irq_enable_r), FUNC(dfruit_state::dfruit_irq_enable_w));
-	map(0xff04, 0xff07).rw(FUNC(dfruit_state::dfruit_ram_bank_r), FUNC(dfruit_state::dfruit_ram_bank_w));
-	map(0xff08, 0xff08).rw(FUNC(dfruit_state::dfruit_rom_bank_r), FUNC(dfruit_state::dfruit_rom_bank_w));
```
Comment (angelosa) [inline src/mame/drivers/dfruit.cpp]: These address maps aided in spotting up the HW for gemcrush, so please reinstate those in both drivers comment out.

Comment (angelosa) [reply src/mame/drivers/dfruit.cpp]: Just the internal non-common address maps (i.e. anything lingering in the 0xfe00-0xffff area) needs to be moved inside the actual map, also use C++ style commentary. If you want to save some typing, any major IDE supports _comment code block_

File: src/devices/machine/tc009xlvc.cpp [hunk 3]
```diff
@@ -14,186 +14,140 @@
...
-WRITE8_MEMBER(tc0091lvc_device::tc0091lvc_paletteram_w)
+void tc0091lvc_device::vram_w(offs_t offset, u8 data)
 {
-	m_palette_ram[offset & 0x1ff] = data;
-
+	/* TODO : offset 0x0000 - 0x3fff is not used?
+	if (offset < 0x4000)
+		return;
+	*/
+
```
Comment (angelosa) [inline src/devices/machine/tc009xlvc.cpp]: This just changes implementation, also again it's not declaring an extra address space.

Comment (cam900) [reply src/devices/machine/tc009xlvc.cpp]: Moved into lw8 in map(0x14000, 0x1ffff).

**Pull request merged**